services/spd/trusty/trusty.c

```diff
@@ -99,6 +99,9 @@ static struct args trusty_context_switch(uint32_t security_state, uint64_t r0,
 	ret.r1 = r1;
 	ret.r0 = r0;
 
+#if CTX_INCLUDE_FPREGS
+	fpregs_context_save(get_fpregs_ctx(cm_get_context(security_state)));
+#endif
```

Comment:
Initially when I looked at this, I wondered why we don't already call `fpregs_context_save()` as part of `cm_el1_sysregs_context_save()`, to avoid you having to do this. However, I know the reason why now...

For general world switch cases (e.g. SMCs, interrupts), this code is the right thing to do, since the specs (e.g. SMC Calling Convention) say FP&SIMD state should be preserved by the firmware. However, this is not the case for PSCI operations that result in a loss of power (e.g. CPU_OFF/CPU_SUSPEND). In those cases, it's expected that the caller preserves FP&SIMD state before invoking the PSCI SMCs. Therefore FP&SIMD save/restore in those case is not necessary, and doing it may have a significant performance impact.

I suggest either you move these FP save/restore invocations to the calling code (except the suspend/resume case), or you inspect the value of r0 here to see if it's not equal to `SMC_FC_CPU_SUSPEND`/`SMC_FC_CPU_RESUME` before saving FP&SIMD state.

Reply:
Thanks for your review and suggestion. The update has been uploaded with changed name info.
I prefer to inspect the value before calling the FP save/restore routine. It seems like that there is no exactly good place to call these rotines  if I move them out from `trusty_context_switch`..